### PR TITLE
Unify segment ZK metadata

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
## Description
Currently there are 3 different classes for segment ZK metadata: `OfflineSegmentZKMetadata`, `RealtimeSegmentZKMetadata`, `LLCRealtimeSegmentZKMetadata`, and 3 different sets of APIs to read them from the ZK.
The problem with this approach:
- Need to know the type of segment before reading the ZK metadata, which is hard to use and makes common code non-sharable
- Impossible to mix different types of ZK metadata. E.g. to support segment push to LLC table

This PR unifies these 3 classes to `SegmentZKMetadata` class which is backed by the `ZNRecord` to:
- Solve the problems above
- Read fields from `ZNRecord` on demand to avoid pre-read all fields (in most cases only one or two fields are required)
- Directly write the fields back to `ZNRecord` to avoid the overhead of creating a new `ZNRecord` when persisting back the segment ZK metadata

The following keys are deprecated, backward-compatibility is maintained:
- `segment.realtime.download.url` and `segment.offline.download.url`: replaced by `segment.download.url`
- `segment.offline.push.time`: replaced by `segment.push.time`
- `segment.offline.refresh.time`: replaced by `segment.refresh.time`
- `segment.name`: use ZNRecord id instead
- `segment.table.name`: decouple segment from table
- `segment.type`: no need to differentiate

## Release Note
Backward-incompatible change: `OfflineSegmentZKMetadata`, `RealtimeSegmentZKMetadata`, `LLCRealtimeSegmentZKMetadata` are removed and replaced with `SegmentZKMetadata`

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java [hunk 1]
```diff
@@ -144,34 +141,25 @@ public static void removeResourceConfigFromPropertyStore(ZkHelixPropertyStore<ZN
...
-        .set(constructPropertyStorePathForSegment(realtimeTableName, realtimeSegmentZKMetadata.getSegmentName()),
-            realtimeSegmentZKMetadata.toZNRecord(), AccessOption.PERSISTENT);
+  public static boolean setSegmentZKMetadata(ZkHelixPropertyStore<ZNRecord> propertyStore, String tableNameWithType,
+      SegmentZKMetadata segmentZKMetadata) {
+    return setSegmentZKMetadata(propertyStore, tableNameWithType, segmentZKMetadata, -1);
   }
 
   @Nullable
-  public static ZNRecord getZnRecord(@Nonnull ZkHelixPropertyStore<ZNRecord> propertyStore, @Nonnull String path) {
+  public static ZNRecord getZnRecord(ZkHelixPropertyStore<ZNRecord> propertyStore, String path) {
```
Comment (klsince) [inline pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java:162]: why removing all @Nonnull annotation?

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java:162]: `Nonnull` and `Nullable` looks quite similar, so we usually only keep `Nullable` and assume the variables are `Nonnull` by default (google convention)

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotRealtimeSegmentManager.java [hunk 2]
```diff
@@ -201,15 +201,13 @@ private synchronized void assignRealtimeSegmentsToServerInstancesIfNecessary() {
           for (String partition : state.getPartitionSet()) {
             // Helix partition is the segment name
             if (SegmentName.isHighLevelConsumerSegmentName(partition)) {
-              HLCSegmentName segName = new HLCSegmentName(partition);
-              RealtimeSegmentZKMetadata realtimeSegmentZKMetadata = ZKMetadataProvider
-                  .getRealtimeSegmentZKMetadata(_pinotHelixResourceManager.getPropertyStore(), segName.getTableName(),
-                      partition);
-              if (realtimeSegmentZKMetadata == null) {
+              SegmentZKMetadata segmentZKMetadata = ZKMetadataProvider
```
Comment (klsince) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotRealtimeSegmentManager.java:204]: looks like segName was not used in the old code. but should it be used actually? segName vs. partition in the getSegmentZKMetadata() call

**Review by klsince — COMMENTED** [review]
finished half; continue on next half.

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java (hunk 1 above)
Comment (klsince) [reply pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java:162]: ic. I see some methods mark one param as Nullable and leave others without annotation. In this case, does compile assume the other Nonnull or Nullable?

Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java:162]: This annotation is mostly for readability purpose, and will be ignored by the compiler. IDE will perform some simple check and warn when invoking methods on `nullable` values.

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentZKMetadata.java [hunk 3]
```diff
@@ -36,209 +34,291 @@
 import org.slf4j.LoggerFactory;
 
 
-public abstract class SegmentZKMetadata implements ZKMetadata {
+public class SegmentZKMetadata implements ZKMetadata {
```
Comment (klsince) [inline pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentZKMetadata.java:37]: looks like changes in this file can be in its own PR to merge firstly? as the interfaces are not changed much.

File: pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManagerTest.java [hunk 4]
```diff
@@ -93,7 +92,8 @@
   private static final long TIMEOUT_IN_MS = 10_000L;
 
   @BeforeClass
-  public void setUp() throws Exception {
+  public void setUp()
```
Comment (klsince) [inline pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManagerTest.java:95]: looks like the changes are mostly on formats. how about open a separate RP to land those first?

File: pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManagerTest.java [hunk 5]
```diff
@@ -90,7 +91,7 @@
   static final long START_TIME_MS = CURRENT_TIME_MS - TimeUnit.HOURS.toMillis(RANDOM.nextInt(24) + 24);
   static final long END_TIME_MS = START_TIME_MS + TimeUnit.HOURS.toMillis(RANDOM.nextInt(24) + 1);
   static final Interval INTERVAL = new Interval(START_TIME_MS, END_TIME_MS);
-  static final String CRC = Long.toString(RANDOM.nextLong());
+  static final String CRC = Long.toString(RANDOM.nextLong() & 0xFFFFFFFFL);
```
Comment (klsince) [inline pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManagerTest.java:94]: curious why `& 0xFFFFFFFFL` here?

File: pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/realtime/SegmentCompletionTest.java [hunk 6]
```diff
@@ -64,7 +64,8 @@
   private final LongMsgOffset s3Offset = new LongMsgOffset(30L);
 
   @BeforeMethod
-  public void testCaseSetup() throws Exception {
+  public void testCaseSetup()
```
Comment (klsince) [inline pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/realtime/SegmentCompletionTest.java:67]: changes in this file are also mainly on formats. The changes are so many that github simply folds the file up. would suggest to open a quick RP to land those format changes firstly.

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/HLRealtimeSegmentDataManager.java [hunk 7]
```diff
@@ -377,10 +377,9 @@ public void run() {
 
           try {
             _segmentLogger.info("Marking current segment as completed in Helix");
-            RealtimeSegmentZKMetadata metadataToOverwrite = new RealtimeSegmentZKMetadata();
+            SegmentZKMetadata metadataToOverwrite = new SegmentZKMetadata(segmentZKMetadata.getSegmentName());
             metadataToOverwrite.setTableName(_tableNameWithType);
-            metadataToOverwrite.setSegmentName(realtimeSegmentZKMetadata.getSegmentName());
-            metadataToOverwrite.setSegmentType(SegmentType.OFFLINE);
```
Comment (klsince) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/HLRealtimeSegmentDataManager.java:383]: hmm.. is this a typo in the old code, or intentionally?

**Review by klsince — COMMENTED** [review]
finished 2nd half. I feel like some changes can be in separate RPs (left comment where I feel so) and get landed firstly to make this one smaller and easier to review. thoughts?

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java [hunk 8]
```diff
@@ -309,89 +272,27 @@ public static Schema getTableSchema(@Nonnull ZkHelixPropertyStore<ZNRecord> prop
...
         if (znRecord != null) {
-          offlineSegmentZKMetadataList.add(new OfflineSegmentZKMetadata(znRecord));
+          segmentsZKMetadata.add(new SegmentZKMetadata(znRecord));
         }
       }
-      int numNullZNRecords = numZNRecords - offlineSegmentZKMetadataList.size();
+      int numNullZNRecords = numZNRecords - segmentsZKMetadata.size();
       if (numNullZNRecords > 0) {
-        LOGGER.warn("Failed to read {}/{} offline segment ZK metadata under path: {}", numZNRecords - numNullZNRecords,
+        LOGGER.warn("Failed to read {}/{} segment ZK metadata under path: {}", numZNRecords - numNullZNRecords,
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/metadata/ZKMetadataProvider.java:292]: Good catch

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentZKMetadata.java [hunk 9]
```diff
@@ -36,209 +34,291 @@
...
+  private Map<String, String> _simpleFields;
 
-  @Deprecated
-  private String _rawTableName;
-
-  public SegmentZKMetadata() {
+  public SegmentZKMetadata(String segmentName) {
+    _znRecord = new ZNRecord(segmentName);
+    _simpleFields = _znRecord.getSimpleFields();
+    // TODO: Remove this field after releasing 0.9.0
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentZKMetadata.java:47]: I think this comment implies that the field is deprecated and we want to keep backward-compatibility

File: pinot-common/src/test/java/org/apache/pinot/common/tier/TierSegmentSelectorTest.java [hunk 10]
```diff
@@ -46,8 +45,8 @@ public void testTimeBasedSegmentSelector() {
     // offline segment
     String segmentName = "segment_0";
     String tableNameWithType = "myTable_OFFLINE";
-    OfflineSegmentZKMetadata offlineSegmentZKMetadata = new OfflineSegmentZKMetadata();
-    offlineSegmentZKMetadata.setSegmentName(segmentName);
+    SegmentZKMetadata offlineSegmentZKMetadata = new SegmentZKMetadata(segmentName);
+    offlineSegmentZKMetadata.setSegmentType(CommonConstants.Segment.SegmentType.OFFLINE);
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-common/src/test/java/org/apache/pinot/common/tier/TierSegmentSelectorTest.java:49]: It is called within the constructor of `OfflineSegmentZKMetadata` and `RealtimeSegmentZKMetadata` which is being cleaned up in this PR. It is already deprecated and will be removed in the next release, so I didn't put it into the constructor so that it can be easily cleaned up

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotRealtimeSegmentManager.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotRealtimeSegmentManager.java:204]: partition is the same as segment name (mentioned in the comment)

File: pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeSegmentValidationManager.java [hunk 11]
```diff
@@ -136,14 +135,14 @@ static long computeRealtimeTotalDocumentInSegments(List<RealtimeSegmentZKMetadat
             groupId = segmentGroupIdName;
           }
           // Discard all segments with different groupids as they are replicas
-          if (groupId.equals(segmentGroupIdName) && realtimeSegmentZKMetadata.getTotalDocs() >= 0) {
-            numTotalDocs += realtimeSegmentZKMetadata.getTotalDocs();
+          if (groupId.equals(segmentGroupIdName) && segmentZKMetadata.getTotalDocs() >= 0) {
+            numTotalDocs += segmentZKMetadata.getTotalDocs();
           }
         }
       } else {
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/validation/RealtimeSegmentValidationManager.java:142]: I think it is needed during the migration from HLC to LLC. Also, out of the scope of this PR

File: pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManagerTest.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManagerTest.java:95]: Sure, will make a PR to only reformat

File: pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManagerTest.java (hunk 5 above)
Comment (Jackie-Jiang) [reply pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManagerTest.java:94]: Added comment: CRC is always non-negative

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/HLRealtimeSegmentDataManager.java (hunk 7 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/HLRealtimeSegmentDataManager.java:383]: Seems a typo to me. It does not matter actually because segment type is already deprecated and is only used to determine the key for download url, which does not apply to the HLC segment.

File: pinot-common/src/main/java/org/apache/pinot/common/tier/TimeBasedTierSegmentSelector.java [hunk 12]
```diff
@@ -55,10 +55,10 @@ public boolean selectSegment(String tableNameWithType, String segmentName) {
     SegmentZKMetadata segmentZKMetadata;
     if (TableNameBuilder.isOfflineTableResource(tableNameWithType)) {
```
Comment (xiangfu0) [inline pinot-common/src/main/java/org/apache/pinot/common/tier/TimeBasedTierSegmentSelector.java:56]: we don't need this if check anymore

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/HLRealtimeSegmentDataManager.java (hunk 7 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/HLRealtimeSegmentDataManager.java:383]: I think we used to claim only consuming segment as real-time segment and all persisted(completed) segments are in `OFFLINE` type.
If we don't have any logic rely on this, then it should be fine. E.g. segment retention.

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentZKMetadata.java [hunk 13]
```diff
@@ -36,209 +34,302 @@
...
-    _segmentType = segmentType;
+    return _znRecord.getId();
   }
 
   public long getStartTimeMs() {
-    if (_startTime > 0 && _timeUnit != null) {
-      return _timeUnit.toMillis(_startTime);
-    } else {
-      return -1;
+    if (_startTimeMs < 0) {
```
Comment (xiangfu0) [inline pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentZKMetadata.java:65]: use <= ?

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentZKMetadata.java [hunk 14]
```diff
@@ -36,209 +34,302 @@
...
-    _startTime = startTime;
+    return _startTimeMs;
   }
 
   public long getEndTimeMs() {
-    if (_endTime > 0 && _timeUnit != null) {
-      return _timeUnit.toMillis(_endTime);
-    } else {
-      return -1;
+    if (_endTimeMs < 0) {
```
Comment (xiangfu0) [inline pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentZKMetadata.java:75]: use <= ?

**Review by xiangfu0 — APPROVED** [review]

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java [hunk 15]
```diff
@@ -1742,37 +1736,36 @@ public ZNRecord getSegmentMetadataZnRecord(String tableNameWithType, String segm
...
     // ZK metadata to refresh the segment (server will compare the segment ZK metadata with the local metadata to decide
     // whether to download the new segment; broker will update the the segment partition info & time boundary based on
     // the segment ZK metadata)
     ZKMetadataUtils
-        .updateSegmentMetadata(offlineSegmentZKMetadata, segmentMetadata, CommonConstants.Segment.SegmentType.OFFLINE);
-    offlineSegmentZKMetadata.setRefreshTime(System.currentTimeMillis());
-    offlineSegmentZKMetadata.setDownloadUrl(downloadUrl);
-    offlineSegmentZKMetadata.setCrypterName(crypter);
-    if (!ZKMetadataProvider.setOfflineSegmentZKMetadata(_propertyStore, offlineTableName, offlineSegmentZKMetadata)) {
+        .updateSegmentMetadata(segmentZKMetadata, segmentMetadata, CommonConstants.Segment.SegmentType.OFFLINE);
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java:1757]: For now, yes. When we remove the segment type completely after releasing `0.9.0`, we can remove it.

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/HLRealtimeSegmentDataManager.java (hunk 7 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/HLRealtimeSegmentDataManager.java:383]: I'll keep it the same as before in case it is used (I don't see any usage though).

File: pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentZKMetadata.java (hunk 13 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentZKMetadata.java:65]: Added a boolean field to track whether the time is cached

**Comment by snleee** [comment]
This will break the external code-base from compilation when they have used `OfflineSegmentZKMetadata, RealtimeSegmentZkMetadata, etc`. One example @ Linkedin would be the purge task generator where we scan the segments. Please add the relevant label (backward-incompat, release-note). 

@jtao15 This change will break the release pipeline. We will need to update our internal code bases to reflect this change. (e.g. `OfflineSegmentZkMetadata -> SegmentZkMetadata`).

**Comment by Jackie-Jiang** [comment]
@snleee Added the labels and the release note. The `SegmentZKMetadata` is not meant to be used by the external code, but the task scheduler might need to access them.

**Comment by jtao15** [comment]
@Jackie-Jiang We have the purge task scheduler in our internal code bases which fetch segment zk metadata from `ClusterInfoAccessor`.
@snleee I'll update our internal code once this pr is merged.

**Comment by jtao15** [comment]
LGTM
